File: dotnet/src/InstrumentComponents/Classes/PowerMeter.cs
using InstrumentComponents.Dialects;
using InstrumentComponents.Errors;
using InstrumentComponents.Identity;
using InstrumentComponents.Kind;
using InstrumentComponents.Session;
using InstrumentComponents.Scpi;

namespace InstrumentComponents.Classes;

public enum PowerUnit
{
    Watt,
    Dbm,
}

public static class PowerUnitExtensions
{
    public static string ScpiName(this PowerUnit unit) => unit switch
    {
        PowerUnit.Watt => "W",
        PowerUnit.Dbm => "DBM",
        _ => "DBM",
    };
}

/// <summary>RF / microwave power meter session view.</summary>
public sealed class PowerMeter : IInstrumentIdentity, IInstrumentShutdown
{
    private readonly InstrumentSession _session;

    public PowerMeter(InstrumentSession session) => _session = session;

    public InstrumentSession Session => _session;

    public Idn QueryIdn() => _session.Idn();

    public void Reset() => _session.Reset();

    /// <summary>Power meters have no output stage; safe-shutdown is a no-op.</summary>
    public void OutputOff()
    {
    }

    private DialectProfile Dialect => _session.DialectFor(InstrumentKind.PowerMeter);

    private string RequireCommand(string key) =>
        Dialect.Command(key) ?? throw new InstrumentUnsupportedException($"power meter dialect missing command '{key}'");

    private string RequireFormatted(string key, params (string Name, string Value)[] vars) =>
        Dialect.FormatCommand(key, vars) ?? throw new InstrumentUnsupportedException($"power meter dialect missing command '{key}'");

    public void ConfigureMeasurement(
        PowerUnit unit,
        bool autoRange,
        bool autoAverage,
        double? correctionFreqHz = null,
        double? offsetDb = null)
    {
        _session.Scpi.Write(RequireFormatted("unit", ("unit", unit.ScpiName())));
        _session.Scpi.Write(RequireFormatted("auto_range", ("state", autoRange ? "ON" : "OFF")));
        _session.Scpi.Write(RequireFormatted("auto_average", ("state", autoAverage ? "ON" : "OFF")));
        if (correctionFreqHz is { } hz)
            _session.Scpi.Write(RequireFormatted("correction_frequency", ("hz", ScpiFormat.Double(hz))));
        if (offsetDb is { } db)
            _session.Scpi.Write(RequireFormatted("offset", ("db", ScpiFormat.Double(db))));
    }

    public void Initiate() => _session.Scpi.Write(RequireCommand("initiate"));

    public double Fetch() => ScpiSession.ParseF64(_session.Scpi.Query(RequireCommand("fetch")));

    public double Read() => ScpiSession.ParseF64(_session.Scpi.Query(RequireCommand("read")));
}

File: crates/instrument/src/classes/power_meter.rs
use super::dialect_io;
use instrument_core::error::Result;
use instrument_core::kind::InstrumentKind;
use instrument_core::scpi::ScpiSession;
use instrument_core::InstrumentSession;

/// Power meter display / readout unit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PowerUnit {
    Watt,
    Dbm,
}

impl PowerUnit {
    pub(crate) fn scpi_name(self) -> &'static str {
        match self {
            Self::Watt => "W",
            Self::Dbm => "DBM",
        }
    }
}

/// RF / microwave power meter session view.
pub struct PowerMeter {
    session: InstrumentSession,
}

impl PowerMeter {
    pub fn new(session: InstrumentSession) -> Self {
        Self { session }
    }

    pub fn session(&self) -> &InstrumentSession {
        &self.session
    }

    pub fn session_mut(&mut self) -> &mut InstrumentSession {
        &mut self.session
    }

    fn dialect(&self) -> &'static instrument_core::DialectProfile {
        self.session.dialect_for(InstrumentKind::PowerMeter)
    }

    fn write_dialect(&mut self, key: &str, vars: &[(&str, String)]) -> Result<()> {
        let cmd = dialect_io::formatted(self.dialect(), key, vars)?;
        self.session.scpi_mut().write(&cmd)
    }

    fn query_dialect(&mut self, key: &str) -> Result<String> {
        let cmd = dialect_io::command(self.dialect(), key)?;
        self.session.scpi_mut().query(cmd)
    }

    /// Configures units, auto-range/average, optional correction frequency and offset.
    pub fn configure_measurement(
        &mut self,
        unit: PowerUnit,
        auto_range: bool,
        auto_average: bool,
        correction_freq_hz: Option<f64>,
        offset_db: Option<f64>,
    ) -> Result<()> {
        let on_off = |enabled: bool| if enabled { "ON" } else { "OFF" };
        self.write_dialect("unit", &[("unit", unit.scpi_name().into())])?;
        self.write_dialect("auto_range", &[("state", on_off(auto_range).into())])?;
        self.write_dialect("auto_average", &[("state", on_off(auto_average).into())])?;
        if let Some(hz) = correction_freq_hz {
            self.write_dialect("correction_frequency", &[("hz", dialect_io::f64_text(hz))])?;
        }
        if let Some(db) = offset_db {
            self.write_dialect("offset", &[("db", dialect_io::f64_text(db))])?;
        }
        Ok(())
    }

    /// Initiates a measurement (INIT).
    pub fn initiate(&mut self) -> Result<()> {
        let cmd = dialect_io::command(self.dialect(), "initiate")?;
        self.session.scpi_mut().write(cmd)
    }

    /// Fetches the last initiated measurement (FETC?).
    pub fn fetch(&mut self) -> Result<f64> {
        ScpiSession::parse_f64(&self.query_dialect("fetch")?)
    }

    /// Reads a measurement immediately (READ?).
    pub fn read(&mut self) -> Result<f64> {
        ScpiSession::parse_f64(&self.query_dialect("read")?)
    }
}
